power_grid_model_c/power_grid_model/include/power_grid_model/math_solver/newton_raphson_se_solver.hpp

```diff
@@ -21,17 +22,443 @@ namespace power_grid_model {
+    GetterType<2, 1> q_P_v() { return this->template get_val<2, 1>(); }
+    GetterType<3, 0> q_Q_theta() { return this->template get_val<3, 0>(); }
+    GetterType<3, 1> q_Q_v() { return this->template get_val<3, 1>(); }
+
+    GetterType<2, 2> r_P_theta() { return this->template get_val<2, 2>(); }
+    GetterType<2, 3> r_P_v() { return this->template get_val<2, 3>(); }
+    GetterType<3, 2> r_Q_theta() { return this->template get_val<3, 2>(); }
+    GetterType<3, 3> r_Q_v() { return this->template get_val<3, 3>(); }
+};
+
 // solver
 template <bool sym> class NewtonRaphsonSESolver {
+
+    struct NRSEJacobian {
+        struct i_side_block {};
+        struct j_side_block {};
+
+        RealTensor<sym> dP_dt{};
+        RealTensor<sym> dP_dv{};
+        RealTensor<sym> dQ_dt{};
+        RealTensor<sym> dQ_dv{};
+    };
+
   public:
-    NewtonRaphsonSESolver(YBus<sym> const& /* y_bus */,
-                          std::shared_ptr<MathModelTopology const> const& /* topo_ptr */) {}
-
-    MathOutput<sym> run_state_estimation(YBus<sym> const& /* y_bus */, StateEstimationInput<sym> const& /* input */,
-                                         double /* err_tol */, Idx /* max_iter */,
-                                         CalculationInfo& /* calculation_info */) {
-        // hide implementation
-        return {};
+    NewtonRaphsonSESolver(YBus<sym> const& y_bus, std::shared_ptr<MathModelTopology const> topo_ptr)
+        : n_bus_{y_bus.size()},
+          math_topo_{std::move(topo_ptr)},
+          data_gain_(y_bus.nnz_lu()),
+          del_x_rhs_(y_bus.size()),
+          x_(y_bus.size()),
+          sparse_solver_{y_bus.shared_indptr_lu(), y_bus.shared_indices_lu(), y_bus.shared_diag_lu()},
+          perm_(y_bus.size()) {}
+
+    MathOutput<sym> run_state_estimation(YBus<sym> const& y_bus, StateEstimationInput<sym> const& input, double err_tol,
+                                         Idx max_iter, CalculationInfo& calculation_info) {
+        // prepare
+        Timer main_timer;
+        Timer sub_timer;
+        MathOutput<sym> output;
+        output.u.resize(n_bus_);
+        output.bus_injection.resize(n_bus_);
+        double max_dev = std::numeric_limits<double>::max();
+
+        main_timer = Timer(calculation_info, 2220, "Math solver");
+
+        // preprocess measured value
+        sub_timer = Timer(calculation_info, 2221, "Pre-process measured value");
+        MeasuredValues<sym> const measured_values{y_bus.shared_topology(), input};
+
+        // initialize voltage with initial angle
+        sub_timer = Timer(calculation_info, 2223, "Initialize voltages");
+        RealValue<sym> const mean_angle_shift = measured_values.mean_angle_shift();
+        for (Idx bus = 0; bus != n_bus_; ++bus) {
+            x_[bus].v() = 1.0;
+            x_[bus].theta() = mean_angle_shift + math_topo_->phase_shift[bus];
+            output.u[bus] = exp(1.0i * x_[bus].theta());
```

Comment:
Do we need to have the complex value `U` here? Shouldn't we just calculate with `V` and `theta`?

Reply:
It seemed like that there would be an extra multiple calculation if it is not stored. Hence it is stored in both forms, just like NR PF solver.
(Maybe it is not so in the new workout. Will keep/remove as per that.)